tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+                if (_isError(_res))                        \
+                    return SLANG_OPTIX_HANDLE_ERROR(_res); \
+            } while (0)
+
+void _optixLogCallback(unsigned int level, const char* tag, const char* message, void* userData)
+{
+    fprintf(stderr, "optix: %s (%s)\n", message, tag);
+}
+
+#    endif
+
+class MemoryCUDAResource : public BufferResource
+{
+public:
+    MemoryCUDAResource(const Desc& _desc)
+        : BufferResource(_desc)
+    {}
+
+    ~MemoryCUDAResource()
+    {
+        if (m_cudaMemory)
+        {
+            SLANG_CUDA_ASSERT_ON_FAIL(cudaFree(m_cudaMemory));
+        }
+    }
+
+    uint64_t getBindlessHandle() { return (uint64_t)m_cudaMemory; }
+
+    void* m_cudaMemory = nullptr;
+};
+
+class TextureCUDAResource : public TextureResource
+{
+public:
+    TextureCUDAResource(const TextureResource::Desc& desc)
+        : TextureResource(desc)
+    {}
+    ~TextureCUDAResource()
+    {
+        if (m_cudaSurfObj)
+        {
+            SLANG_CUDA_ASSERT_ON_FAIL(cuSurfObjectDestroy(m_cudaSurfObj));
+        }
+        if (m_cudaTexObj)
+        {
+            SLANG_CUDA_ASSERT_ON_FAIL(cuTexObjectDestroy(m_cudaTexObj));
+        }
+        if (m_cudaArray)
+        {
+            SLANG_CUDA_ASSERT_ON_FAIL(cuArrayDestroy(m_cudaArray));
+        }
+        if (m_cudaMipMappedArray)
+        {
+            SLANG_CUDA_ASSERT_ON_FAIL(cuMipmappedArrayDestroy(m_cudaMipMappedArray));
+        }
+    }
+
+    uint64_t getBindlessHandle() { return (uint64_t)m_cudaTexObj; }
+
+    // The texObject is for reading 'texture' like things. This is an opaque type, that's backed by
+    // a long long
+    CUtexObject m_cudaTexObj = CUtexObject();
+
+    // The surfObj is for reading/writing 'texture like' things, but not for sampling.
+    CUsurfObject m_cudaSurfObj = CUsurfObject();
+
+    CUarray m_cudaArray = CUarray();
+    CUmipmappedArray m_cudaMipMappedArray = CUmipmappedArray();
+};
+
+class CUDAResourceView : public ResourceView
+{
+public:
+    Desc desc;
+    RefPtr<MemoryCUDAResource> memoryResource = nullptr;
+    RefPtr<TextureCUDAResource> textureResource = nullptr;
+};
+
+class CUDAShaderProgram : public ShaderProgram
+{
+public:
+    CUmodule cudaModule = nullptr;
+    CUfunction cudaKernel;
+    String kernelName;
+    ~CUDAShaderProgram()
+    {
+        if (cudaModule)
+            cuModuleUnload(cudaModule);
+    }
+};
+
+class CUDAPipelineState : public PipelineState
+{
+public:
+    RefPtr<CUDAShaderProgram> shaderProgram;
+};
+
+class CUDAShaderObjectLayout : public ShaderObjectLayout
+{
+public:
+    slang::TypeLayoutReflection* typeLayout = nullptr;
+
+    struct BindingRangeInfo
+    {
+        slang::BindingType bindingType;
+        Index count;
+        Index baseIndex;
+        Index descriptorSetIndex;
+        Index rangeIndexInDescriptorSet;
+        Index uniformOffset;
+        //        Index   subObjectRangeIndex = -1;
+    };
```

Comment:
Scanning ahead, I see that `baseIndex` is that I referred to as `flatIndex` in the previous comment, so that field is the other one that seemingly should stay.

It then seems like `descriptorSetIndex`, `rangeIndexInDescriptorSet`, and `uniformOffset` are *probably* redundant? I guess maybe `uniformOffset` is maybe useful.

(One thing to note is that there is no strong reason to store information in `m_bindingRanges` here that is redundant with what can be queried on `typeLayout`.